File: 프로그래머스/lv2/81302. 거리두기 확인하기/거리두기 확인하기.swift
import Foundation

var arr:[[String]] = [[]]
var check = 1

func solution(_ places:[[String]]) -> [Int] {
    var ans:[Int] = []
    for k in places{
        arr = Array(repeating: Array(repeating: "", count: 5), count: 5)
        for i in 0...k.count-1{
            let a = k[i].map{String($0)}
            for j in 0...4{
                arr[i][j] = a[j]
            }
        }
        check = 1
        for i in 0...4{
            for j in 0...4{
                if(arr[i][j] == "P"){
                    dfs(i,j)
                }
                
            }
        }
        ans.append(check)
        
    }
    return ans
}



func dfs(_ x:Int, _ y:Int){
    let xx = [0,0,-1,1]
    let yy = [1,-1,0,0]
    for k in 0...3{
        let dx = x + xx[k]
        let dy = y + yy[k]
        if(dx >= 0 && dy >= 0 && dx < 5 && dy < 5){
            if(arr[dx][dy] == "P"){
                check = 0
                return
            }
        }
    }
    
    var dx = 0
    var dy = 0
    
    dx = x + 0
    dy = y + 2
    if(dx >= 0 && dy >= 0 && dx < 5 && dy < 5){
        if(arr[dx][dy] == "P"){
            if(arr[dx][dy-1] == "O"){
                check = 0
                return
            }
        }
    }
    dx = x + 0
    dy = y - 2
    if(dx >= 0 && dy >= 0 && dx < 5 && dy < 5){
        if(arr[dx][dy] == "P"){
            if(arr[dx][dy+1] == "O"){
                check = 0
                return
            }
        }
    }
    dx = x + 2
    dy = y + 0
    if(dx >= 0 && dy >= 0 && dx < 5 && dy < 5){
        if(arr[dx][dy] == "P"){
            if(arr[dx-1][dy] == "O"){
                check = 0
                return
            }
        }
    }
    dx = x - 2
    dy = y + 0
    if(dx >= 0 && dy >= 0 && dx < 5 && dy < 5){
        if(arr[dx][dy] == "P"){
            if(arr[dx+1][dy] == "O"){
                check = 0
                return
            }
        }
    }
    
    dx = x + 1
    dy = y + 1
    if(dx >= 0 && dy >= 0 && dx < 5 && dy < 5){
        if(arr[dx][dy] == "P"){
            if(arr[dx-1][dy] == "O" || arr[dx][dy-1] == "O"){
                check = 0
                return
            }
        }
    }
    
    dx = x + 1
    dy = y - 1
    if(dx >= 0 && dy >= 0 && dx < 5 && dy < 5){
        if(arr[dx][dy] == "P"){
            if(arr[dx-1][dy] == "O" || arr[dx][dy+1] == "O"){
                check = 0
                return
            }
        }
    }
    
    dx = x - 1
    dy = y + 1
    if(dx >= 0 && dy >= 0 && dx < 5 && dy < 5){
        if(arr[dx][dy] == "P"){
            if(arr[dx+1][dy] == "O" || arr[dx][dy-1] == "O"){
                check = 0
                return
            }
        }
    }
    
    dx = x - 1
    dy = y - 1
    if(dx >= 0 && dy >= 0 && dx < 5 && dy < 5){
        if(arr[dx][dy] == "P"){
            if(arr[dx+1][dy] == "O" || arr[dx][dy+1] == "O"){
                check = 0
                return
            }
        }
    }
    
}

File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.java
class Solution {
    static int[] ax = {-1,1,0,0};
    static int[] ay = {0,0,1,-1};
    
    public int[] solution(String[][] places) {
        int[] answer = new int[places.length];
        
        for(int i=0;i<places.length;i++){
            String [][] arr = new String[5][5];
            String [] A = places[i];
            for(int j=0;j<5;j++){
                String [] B = A[j].split("");
                for(int k=0;k<5;k++){
                    arr[j][k] = B[k];
                }
            }
            int ans = 1;
            Loop: for(int j=0;j<5;j++){
                for(int k=0;k<5;k++){
                    if(arr[j][k].equals("P")){
                        ans = check(arr,j,k);
                        if(ans == 0){
                            break Loop;
                        }
                    }
                }
            }
            answer[i] = ans;
            
        }
        return answer;
    }
    
    
    public static int check(String[][] arr, int x,int y){
        int ans = 1;
        for(int i=0;i<4;i++){
            int dx = x+ax[i];
            int dy = y+ay[i];
            if(dx>=0 && dy>=0 && dx<5 && dy<5){
                if(arr[dx][dy].equals("P")){
                    ans = 0;
                    return ans;
                }
                else if(arr[dx][dy].equals("O")){
                    ans = doubleCheck(arr,dx,dy,x,y);
                    if(ans == 0){
                        return ans;
                    }
                }
            }
        }
        return ans;
    }
    
    public static int doubleCheck(String [][] arr, int x, int y, int originX, int originY){
        int ans = 1;
        for(int i=0;i<4;i++){
            int dx = x+ax[i];
            int dy = y+ay[i];
            if(dx>=0 && dy>=0 && dx<5 && dy<5){
                if((dx!=originX || dy!=originY) && arr[dx][dy].equals("P")){
                    ans = 0;
                    return ans;
                }
                
            }
        }
        return ans;
    }
}